### PR TITLE
Added workflow as a composite monitor

### PR DESCRIPTION
**Pull request by stevanbz**
### Description
Added workflow
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    override val schedule: Schedule,
+    override val lastUpdateTime: Instant,
+    override val enabledTime: Instant?,
+    // TODO: Check how this behaves during rolling upgrade/multi-version cluster
+    //  Can read/write and parsing break if it's done from an old -> new version of the plugin?
+    val workflowType: WorkflowType,
+    val user: User?,
+    val schemaVersion: Int = NO_SCHEMA_VERSION,
+    val inputs: List<WorkflowInput>,
+    val owner: String? = "alerting"
```
Comment (lezzago) [inline src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt:38]: Can we make this a constant like `DEFAULT_OWNER`?

**Review by lezzago — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/opensearch/commons/alerting/model/Delegate.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * Each underlying monitors defined in the composite monitor sequence input.
+ * They are executed sequentially in the order mentioned.
+ * Optionally accepts chained findings context.
+ * */
+data class Delegate(
+    val order: Int,
+    val monitorId: String,
+    val chainedFindings: ChainedFindings? = null
```
[the remark this answers is not in the document]
Comment (stevanbz) [reply src/main/kotlin/org/opensearch/commons/alerting/model/Delegate.kt:21]: I think no. Each monitor that is currently being processed by the runner will have chainedFindings containing only one monitor - the one that was executed just before. By using the monitor id and the current workflow execution id, we will be able to find the findings (and extract the docIds) - which we can use as a filter for the current monitor execution.

Maybe the property is not named correctly - in general in chainedFindings we are not storing the monitor composition but in inputs property.

File: src/main/kotlin/org/opensearch/commons/alerting/action/GetWorkflowRequest.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.opensearch.common.io.stream.StreamOutput
+import org.opensearch.rest.RestRequest
+import org.opensearch.search.fetch.subphase.FetchSourceContext
+import java.io.IOException
+
+class GetWorkflowRequest : ActionRequest {
+    val workflowId: String
+    val version: Long
+    val method: RestRequest.Method
+    val srcContext: FetchSourceContext?
```
[the remark this answers is not in the document]
Comment (stevanbz) [reply src/main/kotlin/org/opensearch/commons/alerting/action/GetWorkflowRequest.kt:20]: Tbh - it's copy/paste from GetMonitorRequest

File: src/main/kotlin/org/opensearch/commons/alerting/model/Delegate.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (lezzago) [reply src/main/kotlin/org/opensearch/commons/alerting/model/Delegate.kt:21]: Ahh ok, I think we should rename it to `chainedFinding`. Making it plural implies there should be more.

File: src/main/kotlin/org/opensearch/commons/alerting/model/ChainedFindings.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import org.opensearch.common.xcontent.XContentParser
+import org.opensearch.common.xcontent.XContentParserUtils
+import org.opensearch.commons.notifications.model.BaseModel
+import org.opensearch.commons.utils.validateId
+import java.io.IOException
+
+/**
+ * Context passed in delegate monitor to filter data queried by a monitor based on the findings of the given monitor id.
+ */
+data class ChainedFindings(
```
Comment (lezzago) [inline src/main/kotlin/org/opensearch/commons/alerting/model/ChainedFindings.kt:16]: rename to `ChainedFinding`

File: src/main/kotlin/org/opensearch/commons/alerting/AlertingPluginInterface.kt [hunk 5]
```diff
@@ -95,6 +145,30 @@ object AlertingPluginInterface {
         )
     }
 
+    /**
+     * Get Workflow interface.
+     * @param client Node client for making transport action
+     * @param request The request object
+     * @param listener The listener for getting response
+     */
+    fun getWorkflow(
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/AlertingPluginInterface.kt:154]: we can also add execute workflow

File: src/main/kotlin/org/opensearch/commons/alerting/action/DeleteWorkflowRequest.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import org.opensearch.action.ActionRequestValidationException
+import org.opensearch.action.support.WriteRequest
+import org.opensearch.common.io.stream.StreamInput
+import org.opensearch.common.io.stream.StreamOutput
+import java.io.IOException
+
+class DeleteWorkflowRequest : ActionRequest {
+
+    val workflowId: String
+    val deleteDelegateMonitors: Boolean?
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/action/DeleteWorkflowRequest.kt:13]: java docs

File: src/main/kotlin/org/opensearch/commons/alerting/action/DeleteWorkflowRequest.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    val refreshPolicy: WriteRequest.RefreshPolicy
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/action/DeleteWorkflowRequest.kt:14]: remove refresh policy 
we shoudl alway set it as IMMEDIATE and not take input from user

File: src/main/kotlin/org/opensearch/commons/alerting/action/GetWorkflowRequest.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import org.opensearch.action.ActionRequest
+import org.opensearch.action.ActionRequestValidationException
+import org.opensearch.common.io.stream.StreamInput
+import org.opensearch.common.io.stream.StreamOutput
+import org.opensearch.rest.RestRequest
+import java.io.IOException
+
+class GetWorkflowRequest : ActionRequest {
+    val workflowId: String
+    val version: Long
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/action/GetWorkflowRequest.kt:17]: remove version from request. User should be able to fetch by id

File: src/main/kotlin/org/opensearch/commons/alerting/model/ChainedMonitorFindings.kt [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+import org.opensearch.common.xcontent.XContentParser
+import org.opensearch.common.xcontent.XContentParserUtils
+import org.opensearch.commons.notifications.model.BaseModel
+import org.opensearch.commons.utils.validateId
+import java.io.IOException
+
+/**
+ * Context passed in delegate monitor to filter data queried by a monitor based on the findings of the given monitor id.
+ */
+// TODO - Remove the class and move the monitorId to Delegate (as a chainedMonitorId property) if this class won't be updated by adding new properties
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/ChainedMonitorFindings.kt:16]: is the TODO resolved?

File: src/main/kotlin/org/opensearch/commons/alerting/model/Delegate.kt [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+import org.opensearch.commons.utils.validateId
+import java.io.IOException
+
+/**
+ * Each underlying monitors defined in the composite monitor sequence input.
+ * They are executed sequentially in the order mentioned.
+ * Optionally accepts chained findings context.
+ * */
+data class Delegate(
+    val order: Int,
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/Delegate.kt:19]: java docs for field params

File: src/main/kotlin/org/opensearch/commons/alerting/model/Finding.kt [hunk 11]
```diff
@@ -21,9 +21,29 @@ class Finding(
     val monitorName: String,
     val index: String,
     val docLevelQueries: List<DocLevelQuery>,
-    val timestamp: Instant
+    val timestamp: Instant,
+    val executionId: String? = null,
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/Finding.kt:25]: java docs

File: src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+data class Workflow(
+    override val id: String = NO_ID,
+    override val version: Long = NO_VERSION,
+    override val name: String,
+    override val enabled: Boolean,
+    override val schedule: Schedule,
+    override val lastUpdateTime: Instant,
+    override val enabledTime: Instant?,
+    // TODO: Check how this behaves during rolling upgrade/multi-version cluster
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt:32]: is the todo resolved?

**Review by eirsep — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/opensearch/commons/alerting/model/ChainedMonitorFindings.kt (hunk 9 above)
Comment (stevanbz) [reply src/main/kotlin/org/opensearch/commons/alerting/model/ChainedMonitorFindings.kt:16]: No - I just added TODO according to what I discussed with Ashish:
"Yea we can leave it as is, but add a TODO to mention if this class will really be needed in the future if it wont get updated."

File: src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt (hunk 12 above)
Comment (stevanbz) [reply src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt:32]: No

File: src/main/kotlin/org/opensearch/commons/alerting/AlertingPluginInterface.kt (hunk 5 above)
Comment (stevanbz) [reply src/main/kotlin/org/opensearch/commons/alerting/AlertingPluginInterface.kt:154]: We could - but then we need to extract the ExecuteWorkflowRequest. 
Should I add a execute workflow method here or?

File: src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt (hunk 12 above)
Comment (eirsep) [reply src/main/kotlin/org/opensearch/commons/alerting/model/Workflow.kt:32]: that's ok we can test this before merging feature branch into main

**Review by lezzago — APPROVED** [review]

**Pull request merged**